### PR TITLE
added element to show rubric has been edited

### PR DESCRIPTION
**Pull request by kobryan0619**
This PR:

- Cleans up some nits from a previously merged PR
- Modifies the heading of the create/edit page for rubrics to indicate which one a user is on.
- Shows a user that their rubric has been saved after it has been edited. 

Here's an image of the changes:
image

And videos of the interactions:

https://github.com/code-dot-org/code-dot-org/assets/24883357/8ab4d900-91cc-4e7e-b23e-1ad504ec6147

https://github.com/code-dot-org/code-dot-org/assets/24883357/489862f4-910c-4d47-9406-eb9f192c71b6

## Links
- [AITT-163](https://codedotorg.atlassian.net/jira/software/projects/AITT/boards/70?selectedIssue=AITT-163)
- [AITT-169](https://codedotorg.atlassian.net/jira/software/projects/AITT/boards/70?selectedIssue=AITT-169)
- [Magic number nit](https://github.com/code-dot-org/code-dot-org/pull/53666/files#r1322024097)
- [Cleaning up an unclear comment](https://github.com/code-dot-org/code-dot-org/pull/53666/files#r1322036364)

## Testing story

- Modified a test to make sure the headline change was captured.

### CODE REVIEW

File: apps/src/lib/levelbuilder/rubrics/EvidenceDescriptions.jsx
Comment (kobryan0619) [inline apps/src/lib/levelbuilder/rubrics/EvidenceDescriptions.jsx:31]: This is based on feedback from Bethany to get rid of the magic numbers.

File: apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx
Comment (kobryan0619) [inline apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx]: I also added this here to get rid of the magic numbers

File: apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx
Comment (kobryan0619) [inline apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx:159]: This will change depending on if a user is editing or creating a rubric.

**Comment by tess323** [comment]
lgtm

File: apps/src/lib/levelbuilder/rubrics/EvidenceDescriptions.jsx
Comment (bethanyaconnor) [reply apps/src/lib/levelbuilder/rubrics/EvidenceDescriptions.jsx:31]: thank you!

File: apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx
Comment (bethanyaconnor) [inline apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx:196]: Could you also disable the save button when saving?

**Review by bethanyaconnor — APPROVED** [review]
One small ask -- if it's too complex, I don't think it's a big deal. Otherwise looks good to me! Thanks for following up!

File: apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx
Comment (kobryan0619) [inline apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx]: I thought about adding state here for indicating when something should disabled, but since the notification text also provided information about if a user should be able to click the button again, I thought it made more sense just to reuse the `saveNotificationText` state here.

File: apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx
Comment (kobryan0619) [reply apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx:196]: [Done! ](https://github.com/code-dot-org/code-dot-org/pull/53727/files#r1324854664)

File: apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx
Comment (bethanyaconnor) [reply apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx]: don't want to belabor this but it might be more a little more resilient if you made constants for each of the save notification texts then compared to one of those.

So, something like:
```
const SAVING_TEXT = 'Saving...';
const SAVED_TEXT = 'Saved';
// etc

// a lot of JS here

<Button
  disabled={saveNotificationText === SAVING_TEXT}
/>

```

Otherwise, I think using `saveNotificationText` here works fine so no need for the extra state!

Comment (kobryan0619) [reply apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx]: This is actually perfect.  When making this change I kept thinking the `(saveNotificationText === 'Saving...')` is very breakable, but also didn't have a good sense of how to do it better - so thanks for pointing this out!  I made this change.

File: apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx
Comment (wilkie) [inline apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx]: I feel like this is probably true, but I'm used to looking for this. Do we not localize the levelbuilder? It makes sense not to waste translation effort on it, but I was curious.

File: apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js
Comment (wilkie) [inline apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js]: It was tricky to immediately understand why the answer to the length of this should be `3`. What is your test philosophy about this situation? Is it possible to consider:

```
expect(wrapper.find('select#rubric_level_id option')).to.have.length(props.levels.length);
```

The above tells me that it should be specifically responding to the `levels` list in the properties.

This way the test is not dependent on the structure of the default properties, but instead flexible to it. If somebody adds or modifies the default properties, the test will still pass (or it will fail because it now covers a different case... in that case: yay! tests working as they are meant to!)

It's probably not too important to be _this_ robust in tests.. or at least these tests, but this style also lets us potentially randomize the test and create a "fuzzy" case, which I'm a big fan of:

```
// Randomly add a few more levels
props.levels.concat(
  [...Array(random(5,8))].map((_, i) => { return {id: i, name: `level ${i}`}; })
);

const wrapper = mount(<RubricsContainer {...props} />);

// Make sure it handles an arbitrary number of levels
expect(wrapper.find('select#rubric_level_id option')).to.have.length(props.levels.length);
```

Though this is a bit out of the ordinary for our codebase and relies on being proactive about tests that may probabilistically fail. (though this type of flaky test is Good because it Caught Something Wrong!)

File: apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js
Comment (wilkie) [inline apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js:107]: These tests are all pretty aligned to expectations. I generally try to think of extremes... things I might be missing in my manual testing or design. Like, what happens when it has no levels? Is there anything we should test, there?

Also, I'm looking for a test to cover every interaction and state. We do not have a test for the `handleDropdownChange` action nor do we test the enabling/disabling of the save button. I think those would be good candidates for tests. Are there any others?

File: apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx
Comment (bethanyaconnor) [reply apps/src/lib/levelbuilder/rubrics/RubricsContainer.jsx]: Nope, we do not. In fact, we don't let levelbuilders change the language of the site (we had a couple of instances of translated content being saved as the source of truth and it was annoying)

**Review by bethanyaconnor — APPROVED** [review]
LGTM after addressing Wilkie's questions

File: apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js
Comment (kobryan0619) [reply apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js]: Good points.  

This makes me think of at least one time where I added a test to an existing component which required me to add to existing starting data (such as data about levels in this case).  At times this has caused me to break other tests for a variety of reasons which then caused me to edit those tests too.

I can see how this approach (making tests flexible to changing starting data), could help reduce that from happening down the road if someone else were to add to and/or modify the starting data.  I will make the change!  Thanks for walking me through your thinking!

File: apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js
Comment (kobryan0619) [reply apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js:107]: Looking at your candidates:

1. **What happens when it has no levels:** We do have a [ticket](https://codedotorg.atlassian.net/jira/software/projects/AITT/boards/70?selectedIssue=AITT-170) for if there are no levels, a user should never get to this page.  That being said, in a different PR (which will have a test for this), I am planning on making some UI changes to let the user know they need to have a submittable level to be able to have a rubric.
2. **We do not have a test for the `handleDropdownChange` action** - Correct, this also corresponds to a state change.  I created a test for this.  
3. **Test the enabling/disabling of the save button.**  This test was tricky.  I added a test for this (and the other UI changes that happen when hitting the "save" button).  I did some research into the `async` `await` pattern and found a solution.  We use something similar in [two other places in our code base](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/test/unit/code-studio/pd/components/useRegionalPartnerTest.js), but it is unusual.  From what I have seen so far in my time on the engineering side of things, it seems like this case would be covered in an integration test of some type.  At this point we don't have any integration tests for this work flow.  [I created a ticket for it.](https://codedotorg.atlassian.net/jira/software/projects/AITT/boards/70?selectedIssue=AITT-177)
4.  **I'm looking for a test to cover every interaction and state** - I think this PR covers all of the rendered components that are interact-able (The button and the dropdown) and also covers all components that render conditionally (the notified text and the button being disabled).  On the state side it gets a bit tricky.  We certainly cover when the state of `saveNotificationText` or the `selectedLevelForAssessment` is changed.  However, the other state (`learningGoalList`) is largely there so the information can be passed down to other child components.  So the state of this will not change much in this component alone.  Again, for this reason it seems like some integration tests would be helpful.

File: apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js
Comment (bethanyaconnor) [inline apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js]: hm I don't know what this is. Is it from React Testing Library? It's interesting that it works here 🤔

Comment (davidsbailey) [reply apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js]: I had to look it up:

https://legacy.reactjs.org/docs/testing-recipes.html#act

roughly, a helper that makes sure updates have been processed and applied to the DOM before you make any assertions

File: apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js
Comment (davidsbailey) [inline apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js:132]: this is a very creative solution! However we actually have a library which helps us with stubbing out network calls. please look for `sinon.fakeServer` and see if you can copy the pattern we use for that in other unit tests.

File: apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js
Comment (davidsbailey) [inline apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js]: I'd like to be open to new ways of testing things, but after playing with this a bit I am still not sure I understand how this is working. this makes me want to go with a better understood option like `sinon.fakeServer`.

Comment (kobryan0619) [reply apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js]: Thanks for the lead on `fakeServer`!  I will take a look.  The somewhat nested `awaits` were needed to get the test to pass given we are dealing with one event (clicking the save button) triggering 2 different changes on the same component (the text changing from `Saving...` to `saved!`).  It seemed like this was the only way I could figure out for slowing it down to capture both of those items.   

It always makes me uneasy when using something that isn't widely used in our codebase, so the (re)search continues onto `sinon.fakeServer`!

Comment (davidsbailey) [reply apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js]: thanks for the explanation kaitie, I am still not sure I get it because based on your other assertions in the code, the transition to "saving..." seems to be working before you get to the `act` call. `act` does seem like it should be fair game in the future. the two scenarios I'm seeing are:
1. to render a component (this is all of the existing examples in our codebase)
2. to do something like `act(saveButton.simulate('click'))` (apparently not needed here, but that's the idea)

Comment (wilkie) [reply apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js]: I'm suspicious that the `act` call would do anything. I think it coincidentally works to delay the next calls until after the network request logic finishes. I think it might be equivalent to just one or more `await Promise.resolve();` lines, which are basically a type of "yield" to the asynchronous threads?

The ideal thing to do would be to control the network mock in the test so you can catch the request and then have it respond predictably. Which I believe the `sinon.fakeServer` [gives you](https://sinonjs.org/releases/latest/fake-xhr-and-server/).

There, you can spy the request being made... your component code is waiting for the response. Your test then has control over that... so it can call the `respond` method of the network server mock which should synchronously move the component along. I think the test can be written, then, synchronously (without async).

Pseudo-ish-code:

```
server = sinon.fakeServer.create();
server.respondWith("GET", "/some/save/url", [200,
  { "Content-Type": "application/json" },
  '[{ "id": 1024 }]',
]);
button.simulate('click');
server.respond();

// Make sure the button talked to the network
assert(server.requests.length > 0);

// check that the button says "Saved!", etc

// Clean up
server.restore();
```

I would check conventions in other places using `fakeServer` to know what to call it and how to set it up. Probably in a `beforeEach` block.

File: apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js
Comment (wilkie) [reply apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js:107]: I don't think I would use integration tests to test this kind of state. Do we have examples of using integration testing this way? I would probably still use traditional unit testing to test state logic that is passed down to child components. I wrote a looooooong comment that went through and justified this and what I would do, haha, but then I erased it because it was very involved. Basically, I would strive to unit test each component and mock out the state handlers being passed down. This means unit testing the state logic at its source (the RubricsContainer) and testing that it is passed down. Then unit testing the other components such that they properly call them. I don't value integration testing these together (on top of also unit testing them) all that much.

Comment (wilkie) [reply apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js:107]: Happy to share my short novel lol. I'm reflecting and think I might be a bit over-opinionated and surely would be happy with the best solution given the time and value. But I'm happy to pair on this because I think it's tricky for either of us.

**Comment by kobryan0619** [comment]
Alright team!  Thanks again for all the feedback and thoughtfulness here!  Here's where I am at:

1. After Wilke's comments, some of the massive `RubricsContainer.jsx` component was refactored into a helper file.  This file included the `fetch` portion which was located in the `saveRubric` function.
2. New tests were written to test the fetch portion by itself. 
3. The one test that kicked off [this](https://github.com/code-dot-org/code-dot-org/pull/53727#discussion_r1332100608) conversation is still there.  It passes.  It still requires some hand-wavy reasons for being that way.  Unfortunately, because we are using `fetch` instead of `ajax` we cannot use the `fakeServer` lead Dave shared.  But, I learned something along the way there.  I would advocate for leaving this test in because it still does test all of the states changing and the impact on child components.  

I am hoping @wilkie and I can talk through this today or Monday to come to some sort of firm conclusion and get it merged so the other PRs in the queue connected to this can also get in soon.  Of course, if there is anything anyone else here wants to add or raise concerns about, please let me know and hopefully we can pair and/or mob our way through it.

File: apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js
Comment (kobryan0619) [reply apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js]: To somewhat tie up some lose ends, @wilkie was correct that the `act` call was essentially acting as **two** `await Promise.resolve();`.  The best I can tell, this seems to be needed in order for the `Fetch` to be resolved and then for the component to re-render with `Saved!`as the text.  I added a comment in the test to note that.

**Comment by kobryan0619** [comment]
Final follow up to those still watching :) - I chatted with Wilke Friday and identified some additional tests that can be added.  I made [a ticke](https://codedotorg.atlassian.net/browse/AITT-186)t to add those in so I can merge this in.  This will allow us to merge in some other PRs that were waiting for this one to get in.

**Pull request merged**
